<!DOCTYPE html>
<!--==============================================================================
	           "GitHub HTML5 Pandoc Template" v1.2 — by Tristano Ajmone
	==============================================================================
	(c) Tristano Ajmone, 2017, MIT License (MIT). Project's home repository:

	- https://github.com/tajmone/pandoc-goodies

	This template reuses source code taken from the following projects:

	- GitHub Markdown CSS: © Sindre Sorhus, MIT License (MIT):
	  https://github.com/sindresorhus/github-markdown-css

	- Primer CSS: © 2016 GitHub Inc., MIT License (MIT):
	  http://primercss.io/
	==============================================================================-->
<html>
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Yutaka Masuda" />
  <title>Numerical examples from Mrode (2014)</title>
  <style type="text/css">@font-face{font-family:octicons-link;src:url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff')}.markdown-body{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;color:#24292e;font-family:-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";font-size:16px;line-height:1.5;word-wrap:break-word;box-sizing:border-box;min-width:200px;max-width:980px;margin:0 auto;padding:45px}.markdown-body .octicon{display:inline-block;fill:currentColor;vertical-align:text-bottom}.markdown-body a{background-color:transparent;-webkit-text-decoration-skip:objects;color:#0366d6;text-decoration:none}.markdown-body a:active,.markdown-body a:hover{outline-width:0}.markdown-body h1{margin:.67em 0}.markdown-body img{border-style:none}.markdown-body svg:not(:root){overflow:hidden}.markdown-body code,.markdown-body kbd,.markdown-body pre{font-family:monospace,monospace}.markdown-body input{font:inherit;margin:0;overflow:visible;font-family:inherit;font-size:inherit;line-height:inherit}.markdown-body [type=checkbox]{box-sizing:border-box;padding:0}.markdown-body *{box-sizing:border-box}.markdown-body a:hover{text-decoration:underline}.markdown-body strong{font-weight:600}.markdown-body hr{box-sizing:content-box;overflow:hidden;background:0 0;border-bottom:1px solid #dfe2e5}.markdown-body hr::before{display:table;content:""}.markdown-body hr::after{display:table;clear:both;content:""}.markdown-body table{border-spacing:0;border-collapse:collapse;display:block;width:100%;overflow:auto}.markdown-body td,.markdown-body th{padding:0}.markdown-body blockquote{margin:0}.markdown-body ol ol,.markdown-body ul ol{list-style-type:lower-roman}.markdown-body ol ol ol,.markdown-body ol ul ol,.markdown-body ul ol ol,.markdown-body ul ul ol{list-style-type:lower-alpha}.markdown-body dd{margin-left:0}.markdown-body code{font-family:SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace}.markdown-body pre{font:12px SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace;word-wrap:normal}.markdown-body .pl-0{padding-left:0!important}.markdown-body .pl-1{padding-left:4px!important}.markdown-body .pl-2{padding-left:8px!important}.markdown-body .pl-3{padding-left:16px!important}.markdown-body .pl-4{padding-left:24px!important}.markdown-body .pl-5{padding-left:32px!important}.markdown-body .pl-6{padding-left:40px!important}.markdown-body::before{display:table;content:""}.markdown-body::after{display:table;clear:both;content:""}.markdown-body>:first-child{margin-top:0!important}.markdown-body>:last-child{margin-bottom:0!important}.markdown-body a:not([href]){color:inherit;text-decoration:none}.markdown-body .anchor{float:left;padding-right:4px;margin-left:-20px;line-height:1}.markdown-body .anchor:focus{outline:0}.markdown-body blockquote,.markdown-body dl,.markdown-body ol,.markdown-body p,.markdown-body pre,.markdown-body table,.markdown-body ul{margin-top:0;margin-bottom:16px}.markdown-body hr{height:.25em;padding:0;margin:24px 0;background-color:#e1e4e8;border:0}.markdown-body blockquote{padding:0 1em;color:#6a737d;border-left:.25em solid #dfe2e5}.markdown-body blockquote>:first-child{margin-top:0}.markdown-body blockquote>:last-child{margin-bottom:0}.markdown-body kbd{font-size:11px;box-shadow:inset 0 -1px 0 #959da5}.markdown-body h1,.markdown-body h2,.markdown-body h3,.markdown-body h4,.markdown-body h5,.markdown-body h6{margin-top:24px;margin-bottom:16px;font-weight:600;line-height:1.25}.markdown-body h1 .octicon-link,.markdown-body h2 .octicon-link,.markdown-body h3 .octicon-link,.markdown-body h4 .octicon-link,.markdown-body h5 .octicon-link,.markdown-body h6 .octicon-link{color:#1b1f23;vertical-align:middle;visibility:hidden}.markdown-body h1:hover .anchor,.markdown-body h2:hover .anchor,.markdown-body h3:hover .anchor,.markdown-body h4:hover .anchor,.markdown-body h5:hover .anchor,.markdown-body h6:hover .anchor{text-decoration:none}.markdown-body h1:hover .anchor .octicon-link,.markdown-body h2:hover .anchor .octicon-link,.markdown-body h3:hover .anchor .octicon-link,.markdown-body h4:hover .anchor .octicon-link,.markdown-body h5:hover .anchor .octicon-link,.markdown-body h6:hover .anchor .octicon-link{visibility:visible}.markdown-body h1{padding-bottom:.3em;font-size:2em;border-bottom:1px solid #eaecef}.markdown-body h2{padding-bottom:.3em;font-size:1.5em;border-bottom:1px solid #eaecef}.markdown-body h3{font-size:1.25em}.markdown-body h4{font-size:1em}.markdown-body h5{font-size:.875em}.markdown-body h6{font-size:.85em;color:#6a737d}.markdown-body ol,.markdown-body ul{padding-left:2em}.markdown-body ol ol,.markdown-body ol ul,.markdown-body ul ol,.markdown-body ul ul{margin-top:0;margin-bottom:0}.markdown-body li>p{margin-top:16px}.markdown-body li+li{margin-top:.25em}.markdown-body dl{padding:0}.markdown-body dl dt{padding:0;margin-top:16px;font-size:1em;font-style:italic;font-weight:600}.markdown-body dl dd{padding:0 16px;margin-bottom:16px}.markdown-body table th{font-weight:600}.markdown-body table td,.markdown-body table th{padding:6px 13px;border:1px solid #dfe2e5}.markdown-body table tr{background-color:#fff;border-top:1px solid #c6cbd1}.markdown-body table tr:nth-child(2n){background-color:#f6f8fa}.markdown-body img{max-width:100%;box-sizing:content-box;background-color:#fff}.markdown-body code{padding:.2em 0;margin:0;font-size:85%;background-color:rgba(27,31,35,.05);border-radius:3px}.markdown-body code::after,.markdown-body code::before{letter-spacing:-.2em;content:"\00a0"}.markdown-body pre>code{padding:0;margin:0;font-size:100%;word-break:normal;white-space:pre;background:0 0;border:0}.markdown-body .highlight{margin-bottom:16px}.markdown-body .highlight pre{margin-bottom:0;word-break:normal}.markdown-body .highlight pre,.markdown-body pre{padding:16px;overflow:auto;font-size:85%;line-height:1.45;background-color:#f6f8fa;border-radius:3px}.markdown-body pre code{display:inline;max-width:auto;padding:0;margin:0;overflow:visible;line-height:inherit;word-wrap:normal;background-color:transparent;border:0}.markdown-body pre code::after,.markdown-body pre code::before{content:normal}.markdown-body .full-commit .btn-outline:not(:disabled):hover{color:#005cc5;border-color:#005cc5}.markdown-body kbd{display:inline-block;padding:3px 5px;font:11px SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace;line-height:10px;color:#444d56;vertical-align:middle;background-color:#fcfcfc;border:1px solid #c6cbd1;border-bottom-color:#959da5;border-radius:3px;box-shadow:inset 0 -1px 0 #959da5}.markdown-body :checked+.radio-label{position:relative;z-index:1;border-color:#0366d6}.markdown-body .task-list-item{list-style-type:none}.markdown-body .task-list-item+.task-list-item{margin-top:3px}.markdown-body .task-list-item input{margin:0 .2em .25em -1.6em;vertical-align:middle}.markdown-body hr{border-bottom-color:#eee}.flash{position:relative;padding:16px;color:#246;background-color:#e2eef9;border:1px solid #bac6d3;border-radius:3px}.flash p:last-child{margin-bottom:0}.flash-messages{margin-bottom:24px}.flash-warn{color:#4c4a42;background-color:#fff9ea;border-color:#dfd8c2}.flash-error{color:#911;background-color:#fcdede;border-color:#d2b2b2}.flash-success{color:#22662c;background-color:#e2f9e5;border-color:#bad3be}.flash-plain{color:#4c4a42;background-color:#f5f5f5;border-color:#c1c1c1}</style>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">pre[data-language='renumf90']{background-color:#EFEFFE;}</style>
  <style type="text/css">pre[data-language='blupf90']{background-color:#E7FFE7;}</style>
  <style type="text/css">pre[data-language='output']{background-color:#FFFFFF;border:thin solid black;}</style>
  <style type="text/css">pre[data-language='shell']{background-color:#FFFFFF;border:thin dashed;}</style>
  <style type="text/css">pre[data-caption]::before{content:attr(data-caption) "\a\a";font-size:large;font-weight:bold;border-bottom:thin solid black}</style>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
  </script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<article class="markdown-body">
<header>
<h1 class="title">Numerical examples from Mrode (2014)</h1>
<p class="author">Yutaka Masuda</p>
<p class="date">September 2019</p>
</header>
<nav id="TOC">
<ul>
<li><a href="#gblup">GBLUP</a><ul>
<li><a href="#model">Model</a></li>
<li><a href="#common-files">Common files</a></li>
<li><a href="#the-first-approach-with-a-text-file-for-g-inverse">The first approach with a text file for G-inverse</a></li>
<li><a href="#second-approach-with-a-binary-file-for-g-inverse">Second approach with a binary file for G-inverse</a></li>
</ul></li>
</ul>
</nav>
<p>
Back to <a href="./index.html">index.html</a>.
</p>
<h1 id="gblup">GBLUP</h1>
<h2 id="model">Model</h2>
<p>The SNP marker model is equivalent to a model, so-called GBLUP, under some assumptions. The model includes a genomic relationship matrix, <span class="math inline">\(\mathbf{G}\)</span>. This matrix is calculated using SNP marker data. It is analog to the pedigree-based additive relationship matrix although there is an essential difference between the two matrices; see VanRaden (2008) for details. The <span class="math inline">\(\mathbf{G}\)</span> matrix may be defined in several ways but the best-known one is defined as <span class="math display">\[
\mathbf{G}=\frac{\mathbf{ZZ}&#39;}{2\sum_j p_j(1-p_j)}
\]</span> where <span class="math inline">\(\mathbf{Z}\)</span> contains genotypes adjusted with allele frequency and <span class="math inline">\(p_j\)</span> is the allele frequency for marker <span class="math inline">\(j\)</span> (VanRaden 2008). When the number of genotypes is limited, the above <span class="math inline">\(\mathbf{G}\)</span> is most likely singular, and its inverse does not exist. To avoid this issue, we blend the above <span class="math inline">\(\mathbf{G}\)</span> with a numerator relationship matrix for genotyped animals (<span class="math inline">\(\mathbf{A}_{22}\)</span>). Therefore, as long as we use this blending, the pedigree file is still needed. The formula of the blending is <span class="math display">\[
\mathbf{G}_w = w\mathbf{G}+(1-w)\mathbf{A}_{22}
\]</span> where <span class="math inline">\(\mathbf{G}_w\)</span> is the final (weighted) genomic relationship matrix and <span class="math inline">\(w\)</span> is a weight. When <span class="math inline">\(\mathbf{A}_{22}\)</span> is not available, <span class="math inline">\(\mathbf{I}\)</span> may be useful. An appropriate weight usually ranges from 0.80 to 0.99 depending on the profile of <span class="math inline">\(\mathbf{G}\)</span>. This blending is expected to makes G be positive definite. The weighted matrix, <span class="math inline">\(\mathbf{G}_{w}\)</span>, will be used as a genomic relationship matrix in prediction.</p>
<p>The system of mixed model equations is <span class="math display">\[
\left[
\begin{array}{ll}
\mathbf{X}&#39;\mathbf{R}^{-1}\mathbf{X} &amp; \mathbf{X}&#39;\mathbf{R}^{-1}\mathbf{W}\\
\mathbf{W}&#39;\mathbf{R}^{-1}\mathbf{X} &amp; \mathbf{W}&#39;\mathbf{R}^{-1}\mathbf{W}+\mathbf{G}_{w}/\sigma_u^2
\end{array}
\right]
\left[
\begin{array}{c}
\mathbf{\hat{b}}\\
\mathbf{\hat{u}}
\end{array}
\right]
=
\left[
\begin{array}{l}
\mathbf{X}&#39;\mathbf{R}^{-1}\mathbf{y} \\
\mathbf{W}&#39;\mathbf{R}^{-1}\mathbf{y} \\
\end{array}
\right]
\]</span> and <span class="math inline">\(\sigma_u^2 = 35.241\)</span> and <span class="math inline">\(\sigma_e^2 = 245.0\)</span>. In the simplest case, only the general mean (<span class="math inline">\(\mu\)</span>) is considered as the fixed effect, and <span class="math inline">\(\hat{\mathbf{b}}\)</span> reduces to <span class="math inline">\(\hat{\mu}\)</span>.</p>
<p>BLUPF90 is designed for the single-step GBLUP model which has the inverse of a numerator relationship matrix (<span class="math inline">\(\mathbf{A}^{-1}\)</span>). It means that the final equations should include <span class="math inline">\(\mathbf{A}^{-1}\)</span>, which is not usually used in GBLUP. There are several ways to perform GBLUP with BLUPF90. Here I introduce you to a method that has the following 2 steps: 1) preparation of <span class="math inline">\(\mathbf{G}^{-1}\)</span> and 2) build and solve the equations.</p>
<p>In the former step, we can use the PREGSF90 program. This program accepts the same parameter file as BLUPF90. It calculates <span class="math inline">\(\mathbf{G}^{-1}\)</span> with various options and saves it to a file. BLUPF90 can read the file and use it to build the mixed model equations. BLUPF90 supports two file-formats for <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span>: a text file and a binary file. The use of the text format is simpler especially for a small data set and easier to understand how to run GBLUP using BLUPF90. The use of the binary file is more efficient but you need a trick which mimics single-step GBLUP removing the pedigree relationship matrix. In this section, first, we will use the text format to demonstrate how <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span> can be calculated with PREGSF90 with an educational parameter file. Then, we move to the second approach using the binary file with a different parameter file.</p>
<p>Note that the method introduced here may seem to be complicated. This is because we will not use RENUMF90 here. When using the renumbering tool, GBLUP will be performed easily. In the later chapter, we will see how RENUMF90 simply prepares the files required in pure GBLUP.</p>
<h2 id="common-files">Common files</h2>
<p>In both approaches, we will use the common marker file and the cross-reference file. The SNP file is as follows.</p>
<pre data-language="text" data-caption="snp_mr11c.txt"><code>13 20110002120000000000000000000000000000000000000000
...
26 10110201000000000000000000000000000000000000000000</code></pre>
<p>The format of the SNP file is described in the previous chapter. The textbook uses only 10 SNP markers but the above file contains 50 markers. The extra 40 markers are padding because BLUPF90 can’t read fewer than 50 markers. All animals have the genotype 0 for the extra markers so the markers don’t provide any information. The resulting <span class="math inline">\(\mathbf{G}\)</span> is the same as using only the first 10 markers.</p>
<p>The accompanying cross-reference file is also needed. This file relates the ID for genotyped animals and the order of animals in the marker file. The detailed explanation is also available in the previous chapter.</p>
<pre data-language="text" data-caption="snp_mr11c_XrefID.txt"><code>13 13
14 14
15 15
16 16
17 17
18 18
19 19
20 20
21 21
22 22
23 23
24 24
25 25
26 26</code></pre>
<h2 id="the-first-approach-with-a-text-file-for-g-inverse">The first approach with a text file for G-inverse</h2>
<h3 id="for-pregsf90">For PREGSF90</h3>
<p>To obtain the text file to have <span class="math inline">\(\mathbf{G}^{-1}_{w}\)</span> using PREGSF90, we use the data file (<code>data_mr11c1.txt</code>) as before except for the last column for genotyped animals with different coding. Note that the data file is not really read by PREGSF90 but we need it, because the program checks whether the data file is present or not.</p>
<pre data-language="text" data-caption="data_mr11c1.txt"><code> 13  0  0  1 558  9.0  0.00179211 1
...
 20 14  9  1 103  4.8  0.00970874 8</code></pre>
<ol type="1">
<li>The original animal ID</li>
<li>The original sire ID</li>
<li>The original dam ID</li>
<li>general mean</li>
<li>EDC</li>
<li>Observation (Fat DYD)</li>
<li>1/EDC</li>
<li>new ID for genotyped animals (in the order of the marker file - <em>this is important</em>)</li>
</ol>
<p>The key is the 8-th column. This approach needs a new ID corresponding to the order in the marker file. For example, animal 13 appears in the 1st row in the marker file, and it has 1 as a new ID. Later, we will see the details of the new ID.</p>
<p>The pedigree file (<code>pedigree_mr11c1.txt</code>) contains all the animals with and without genotypes. The program creates a subset of relationship matrix (<span class="math inline">\(\mathbf{A}_{22}\)</span>) only for genotyped animals.</p>
<p>The parameter file for PREGSF90 is the same as BLUPF90.</p>
<pre data-language="blupf90" data-caption="preparam_mr11c1.txt"><code>#
# Approach 1: creates a text G - Inverse file with PREGSF90
#
DATAFILE
data_mr11c1.txt
NUMBER_OF_TRAITS
1
NUMBER_OF_EFFECTS
2
OBSERVATION(S)
6
WEIGHT(S)
7
EFFECTS:
4  1 cross
1 26 cross
RANDOM_RESIDUAL VALUES
245.0
RANDOM_GROUP
2
RANDOM_TYPE
add_animal
FILE
pedigree_mr11c1.txt
(CO)VARIANCES
35.25
OPTION SNP_file snp_mr11c.txt snp_mr11c_XrefID.txt
OPTION no_quality_control
OPTION AlphaBeta 0.99 0.01
OPTION tunedG 0
OPTION saveAscii
OPTION saveG
OPTION saveGInverse
OPTION createGimA22i 0</code></pre>
<p>The purpose of PREGSF90 is to create a text file with the elements of <span class="math inline">\(\mathbf{G}_{w}\)</span>. By default, PREGSF90 tries to manipulate <span class="math inline">\(\mathbf{G}_{w}\)</span> in many ways. The options change the behavior of PREGSF90. Note that the options are case-sensitive. Simple explanations for the options are available:</p>
<ul>
<li><code>SNP_file</code> = Invoking the genomic routine. It needs 2 values in this case: the name of the marker file and the name of the cross-reference file.</li>
<li><code>no_quality_control</code> = Turns off the quality control process to remove unqualified markers and genotyped animals. In this example, we will use all the markers and genotyped animals.</li>
<li><code>AlphaBeta</code> = Defines weight <span class="math inline">\(w\)</span>. This option has 2 values: <span class="math inline">\(\alpha=w\)</span> and <span class="math inline">\(\beta=1-w\)</span>. In this example, we put <span class="math inline">\(w = 0.99\)</span> which is implicitly used by the author.</li>
<li><code>tunedG</code> = Controls the tuning method for <span class="math inline">\(\mathbf{G}_w\)</span>. It tries to adjust the scale of <span class="math inline">\(\mathbf{G}\)</span> to <span class="math inline">\(\mathbf{A}_{22}\)</span> using a method described by Chen et al. (2011). This tuning is <em>a different step as blending</em>. The argument 0 turns off the adjustment (i.e. uses <span class="math inline">\(\mathbf{G}_{w}\)</span> ).</li>
<li><code>saveAscii</code> = Uses the text format for a file to save <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span>. Without this option, the program saves the matrix to a binary file (a machine-readable file).</li>
<li><code>saveG</code> = Saves <span class="math inline">\(\mathbf{G}_w\)</span> to a file. The file name will be <code>G</code>. Actually, you don’t have to put this option in this analysis. We just use it as an educational purpose to see the actual values in <span class="math inline">\(\mathbf{G}_{w}\)</span>.</li>
<li><code>saveGInverse</code> = Saves <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span> to a file. The file name will be <code>Gi</code>.</li>
<li><code>createGimA22i</code> = Turns off the calculation of <span class="math inline">\(\mathbf{G}_{w}^{-1}-\mathbf{A}_{22}\)</span> which is required in single-step GBLUP. You don’t have to put this option in this small example because the computing cost is negligible. Turning off this option is important for a large-size problem in GBLUP.</li>
</ul>
<p>We have more than 100 options just for PREGSF90. Many of them are implemented for experimental purposes and unused. We will demonstrate such options in the later chapter. In this section, a user can just recognize that there are plenty of options for PREGSF90, but only a few options may be essential for the regular use in genomic prediction.</p>
<p>You can run PREGSF90 with the above parameter file. You can see the basic statistics for <span class="math inline">\(\mathbf{G}_{w}\)</span> on the screen. Also, you will find that some files are created.</p>
<ul>
<li><code>Gen_call_rate</code> = Empty in this case. Just ignore it.</li>
<li><code>freq.count</code> = Allele frequency for each marker. The values should be the same as the textbook (see p.181).</li>
<li><code>sum2pq</code> = <span class="math inline">\(2\sum_j p_j(1-p_j)\)</span></li>
<li><code>G</code> = Elements in <span class="math inline">\(\mathbf{G}_w\)</span>. The values are slightly different from the textbook (p.186) because the textbook values are from the pure <span class="math inline">\(\mathbf{G}\)</span> (not blended with <span class="math inline">\(\mathbf{A}_{22}\)</span>). You can find the same values as the textbook when you put <code>OPTION AlphaBeta 1.0 0.0</code>, but the program cannot calculate its inverse because it is singular.</li>
<li><code>Gi</code> = Elements in <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span>. Only the diagonals and the upper triangular elements are stored.</li>
</ul>
<p>Among the above files, only <code>Gi</code> is required in the analysis with BLUPF90. The first 5 rows and the last 5 rows in this file are as follows.</p>
<pre><code>1 1 36.374112147756
1 2 -27.980861774302
2 2 88.270926666258
1 3 -8.255548353247
2 3 31.300275470378
...
10 14 2.733882533018
11 14 6.555679888597
12 14 5.120296531692
13 14 4.504947248223
14 14 6.133181220876</code></pre>
<p>The first 2 integers are the row and column indices for the element. Note that the maximum row and column ID is 14 because the rank of this matrix (<span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span>) is 14. So the row/column index doesn’t directly refer to the animal’s ID. PREGSF90 internally re-assigns IDs for genotyped animals as the consecutive order of animals in the marker file. For example, row 1 and column 1 (the first line in the above file) corresponds to the relationship between animal 13 and 13 (i.e., the diagonal for the animal 13) because the animal 13 appears the first in the marker file (animal 13 = new ID 1). The row 1 and column 2 (the second line) is for the relationships between animal 13 and 14 because the animal 14 appears secondly in the marker file (animal 14 = new ID 2). Therefore, we should use the new ID to refer to genotyped animals.</p>
<p>We have already prepared the data file with the new ID (the 8th column). BLUPF90 can read the new column as an animal ID.</p>
<h3 id="for-blupf90">For BLUPF90</h3>
<p>The parameter file for BLUPF90 is shown below.</p>
<pre data-language="blupf90" data-caption="param_mr11c1.txt"><code>DATAFILE
data_mr11c1.txt
NUMBER_OF_TRAITS
1
NUMBER_OF_EFFECTS
2
OBSERVATION(S)
6
WEIGHT(S)

EFFECTS:
4  1 cross
8 14 cross  # new ID (renumbered only for genotyped animals)
RANDOM_RESIDUAL VALUES
245.0
RANDOM_GROUP
2
RANDOM_TYPE
user_file
FILE
Gi
(CO)VARIANCES
35.25
OPTION solv_method FSPAK</code></pre>
<p>We use the keyword <code>user_file</code> to read the text file for <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span> With this keyword, we don’t need the marker file anymore. BLUPF90 simply read the file and put the elements into the equations. Run BLUPF90 with the parameter file, and you will find the results. The solutions look very similar to the values in the textbook (p.185; GBLUP).</p>
<h2 id="second-approach-with-a-binary-file-for-g-inverse">Second approach with a binary file for G-inverse</h2>
<p>The parameter file for PREGSF90 is very similar to the previous one.</p>
<pre data-language="blupf90" data-caption="preparam_mr11c2.txt"><code>DATAFILE
data_mr11c1.txt
NUMBER_OF_TRAITS
1
NUMBER_OF_EFFECTS
2
OBSERVATION(S)
6
WEIGHT(S)

EFFECTS:
4  1 cross
1 26 cross
RANDOM_RESIDUAL VALUES
245.0
RANDOM_GROUP
2
RANDOM_TYPE
add_animal
FILE
pedigree_mr11c1.txt
(CO)VARIANCES
35.25
OPTION SNP_file snp_mr11c.txt snp_mr11c_XrefID.txt
OPTION no_quality_control
OPTION AlphaBeta 0.99 0.01
OPTION tunedG 0
#OPTION saveAscii
#OPTION saveG
OPTION saveGInverse
OPTION createGimA22i 0</code></pre>
<p>The only differences between this parameter file and the previous one are options. We remove (actually comment out) the two options. With this parameter file, the G-inverse file (<code>Gi</code>) is saved as a binary format readable only for the computer. Also, we don’t need a file, <code>G</code>. Run PREGSF90 with this parameter file, and you can see the very similar output on the screen.</p>
<p>With this approach, we trick BLUPF90 as if the single-step GBLUP was applied. In this analysis, we shouldn’t use the pedigree file. The prepared <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span> is already blended with <span class="math inline">\(\mathbf{A}_{22}\)</span> so we do not need the pedigree file. We just prepare a dummy, empry pedigree file. This file contains 26 lines, which is the same number of lines as the original pedigree.</p>
<pre data-language="text" data-caption="pedigree_mr11c2.txt"><code>0 0 0
...
0 0 0</code></pre>
<p>We need the data, (empty) pedigree, marker and cross-reference files as well as <code>Gi</code> for the analysis with BLUPF90. The parameter file for BLUPF90 is slightly different from one for PREGSF90.</p>
<pre data-language="blupf90" data-caption="param_mr11c2.txt"><code>DATAFILE
data_mr11c1.txt
NUMBER_OF_TRAITS
1
NUMBER_OF_EFFECTS
2
OBSERVATION(S)
6
WEIGHT(S)

EFFECTS:
4  1 cross
1 26 cross  # use the original ID
RANDOM_RESIDUAL VALUES
245.0
RANDOM_GROUP
2
RANDOM_TYPE
add_animal
FILE
pedigree_mr11c2.txt
(CO)VARIANCES
35.25
OPTION SNP_file snp_mr11c.txt snp_mr11c_XrefID.txt
OPTION readGimA22i Gi
OPTION solv_method FSPAK</code></pre>
<p>This parameter file is for single-step GBLUP. In this case, the pedigree file is empty, and <span class="math inline">\(\mathbf{A}^{-1}=\mathbf{0}\)</span>. The program will read <code>Gi</code> instead of creating <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span> and <span class="math inline">\(\mathbf{A}_{22}\)</span> with the second option (<code>OPTION readGimA22i</code>). The program reads <span class="math inline">\(\mathbf{G}_{w}^{-1}\)</span>, and it performs GBLUP.</p>
<p>Run BLUPF90 with the parameter file, and you can find the solutions that are identical to approach 1. You can see many warnings (<code>Animal Id equal to Sire/Dam Id - relationships removed !!</code>) because we use a trick which never happens in the regular single-step GBLUP.</p>
<p>Again, the approaches demonstrated here require some manual operations. In the later chapter, we will see an automatic approach to perform GBLUP using RENUMF90 and BLUPF90.</p>
<p>
Back to <a href="./index.html">index.html</a>.
</p>
</article>
</body>
</html>
