<!DOCTYPE html>
<!--==============================================================================
	           "GitHub HTML5 Pandoc Template" v1.2 — by Tristano Ajmone
	==============================================================================
	(c) Tristano Ajmone, 2017, MIT License (MIT). Project's home repository:

	- https://github.com/tajmone/pandoc-goodies

	This template reuses source code taken from the following projects:

	- GitHub Markdown CSS: © Sindre Sorhus, MIT License (MIT):
	  https://github.com/sindresorhus/github-markdown-css

	- Primer CSS: © 2016 GitHub Inc., MIT License (MIT):
	  http://primercss.io/
	==============================================================================-->
<html>
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Yutaka Masuda" />
  <title>Data preparation with RENUMF90</title>
  <style type="text/css">@font-face{font-family:octicons-link;src:url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff')}.markdown-body{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;color:#24292e;font-family:-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";font-size:16px;line-height:1.5;word-wrap:break-word;box-sizing:border-box;min-width:200px;max-width:980px;margin:0 auto;padding:45px}.markdown-body .octicon{display:inline-block;fill:currentColor;vertical-align:text-bottom}.markdown-body a{background-color:transparent;-webkit-text-decoration-skip:objects;color:#0366d6;text-decoration:none}.markdown-body a:active,.markdown-body a:hover{outline-width:0}.markdown-body h1{margin:.67em 0}.markdown-body img{border-style:none}.markdown-body svg:not(:root){overflow:hidden}.markdown-body code,.markdown-body kbd,.markdown-body pre{font-family:monospace,monospace}.markdown-body input{font:inherit;margin:0;overflow:visible;font-family:inherit;font-size:inherit;line-height:inherit}.markdown-body [type=checkbox]{box-sizing:border-box;padding:0}.markdown-body *{box-sizing:border-box}.markdown-body a:hover{text-decoration:underline}.markdown-body strong{font-weight:600}.markdown-body hr{box-sizing:content-box;overflow:hidden;background:0 0;border-bottom:1px solid #dfe2e5}.markdown-body hr::before{display:table;content:""}.markdown-body hr::after{display:table;clear:both;content:""}.markdown-body table{border-spacing:0;border-collapse:collapse;display:block;width:100%;overflow:auto}.markdown-body td,.markdown-body th{padding:0}.markdown-body blockquote{margin:0}.markdown-body ol ol,.markdown-body ul ol{list-style-type:lower-roman}.markdown-body ol ol ol,.markdown-body ol ul ol,.markdown-body ul ol ol,.markdown-body ul ul ol{list-style-type:lower-alpha}.markdown-body dd{margin-left:0}.markdown-body code{font-family:SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace}.markdown-body pre{font:12px SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace;word-wrap:normal}.markdown-body .pl-0{padding-left:0!important}.markdown-body .pl-1{padding-left:4px!important}.markdown-body .pl-2{padding-left:8px!important}.markdown-body .pl-3{padding-left:16px!important}.markdown-body .pl-4{padding-left:24px!important}.markdown-body .pl-5{padding-left:32px!important}.markdown-body .pl-6{padding-left:40px!important}.markdown-body::before{display:table;content:""}.markdown-body::after{display:table;clear:both;content:""}.markdown-body>:first-child{margin-top:0!important}.markdown-body>:last-child{margin-bottom:0!important}.markdown-body a:not([href]){color:inherit;text-decoration:none}.markdown-body .anchor{float:left;padding-right:4px;margin-left:-20px;line-height:1}.markdown-body .anchor:focus{outline:0}.markdown-body blockquote,.markdown-body dl,.markdown-body ol,.markdown-body p,.markdown-body pre,.markdown-body table,.markdown-body ul{margin-top:0;margin-bottom:16px}.markdown-body hr{height:.25em;padding:0;margin:24px 0;background-color:#e1e4e8;border:0}.markdown-body blockquote{padding:0 1em;color:#6a737d;border-left:.25em solid #dfe2e5}.markdown-body blockquote>:first-child{margin-top:0}.markdown-body blockquote>:last-child{margin-bottom:0}.markdown-body kbd{font-size:11px;box-shadow:inset 0 -1px 0 #959da5}.markdown-body h1,.markdown-body h2,.markdown-body h3,.markdown-body h4,.markdown-body h5,.markdown-body h6{margin-top:24px;margin-bottom:16px;font-weight:600;line-height:1.25}.markdown-body h1 .octicon-link,.markdown-body h2 .octicon-link,.markdown-body h3 .octicon-link,.markdown-body h4 .octicon-link,.markdown-body h5 .octicon-link,.markdown-body h6 .octicon-link{color:#1b1f23;vertical-align:middle;visibility:hidden}.markdown-body h1:hover .anchor,.markdown-body h2:hover .anchor,.markdown-body h3:hover .anchor,.markdown-body h4:hover .anchor,.markdown-body h5:hover .anchor,.markdown-body h6:hover .anchor{text-decoration:none}.markdown-body h1:hover .anchor .octicon-link,.markdown-body h2:hover .anchor .octicon-link,.markdown-body h3:hover .anchor .octicon-link,.markdown-body h4:hover .anchor .octicon-link,.markdown-body h5:hover .anchor .octicon-link,.markdown-body h6:hover .anchor .octicon-link{visibility:visible}.markdown-body h1{padding-bottom:.3em;font-size:2em;border-bottom:1px solid #eaecef}.markdown-body h2{padding-bottom:.3em;font-size:1.5em;border-bottom:1px solid #eaecef}.markdown-body h3{font-size:1.25em}.markdown-body h4{font-size:1em}.markdown-body h5{font-size:.875em}.markdown-body h6{font-size:.85em;color:#6a737d}.markdown-body ol,.markdown-body ul{padding-left:2em}.markdown-body ol ol,.markdown-body ol ul,.markdown-body ul ol,.markdown-body ul ul{margin-top:0;margin-bottom:0}.markdown-body li>p{margin-top:16px}.markdown-body li+li{margin-top:.25em}.markdown-body dl{padding:0}.markdown-body dl dt{padding:0;margin-top:16px;font-size:1em;font-style:italic;font-weight:600}.markdown-body dl dd{padding:0 16px;margin-bottom:16px}.markdown-body table th{font-weight:600}.markdown-body table td,.markdown-body table th{padding:6px 13px;border:1px solid #dfe2e5}.markdown-body table tr{background-color:#fff;border-top:1px solid #c6cbd1}.markdown-body table tr:nth-child(2n){background-color:#f6f8fa}.markdown-body img{max-width:100%;box-sizing:content-box;background-color:#fff}.markdown-body code{padding:.2em 0;margin:0;font-size:85%;background-color:rgba(27,31,35,.05);border-radius:3px}.markdown-body code::after,.markdown-body code::before{letter-spacing:-.2em;content:"\00a0"}.markdown-body pre>code{padding:0;margin:0;font-size:100%;word-break:normal;white-space:pre;background:0 0;border:0}.markdown-body .highlight{margin-bottom:16px}.markdown-body .highlight pre{margin-bottom:0;word-break:normal}.markdown-body .highlight pre,.markdown-body pre{padding:16px;overflow:auto;font-size:85%;line-height:1.45;background-color:#f6f8fa;border-radius:3px}.markdown-body pre code{display:inline;max-width:auto;padding:0;margin:0;overflow:visible;line-height:inherit;word-wrap:normal;background-color:transparent;border:0}.markdown-body pre code::after,.markdown-body pre code::before{content:normal}.markdown-body .full-commit .btn-outline:not(:disabled):hover{color:#005cc5;border-color:#005cc5}.markdown-body kbd{display:inline-block;padding:3px 5px;font:11px SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace;line-height:10px;color:#444d56;vertical-align:middle;background-color:#fcfcfc;border:1px solid #c6cbd1;border-bottom-color:#959da5;border-radius:3px;box-shadow:inset 0 -1px 0 #959da5}.markdown-body :checked+.radio-label{position:relative;z-index:1;border-color:#0366d6}.markdown-body .task-list-item{list-style-type:none}.markdown-body .task-list-item+.task-list-item{margin-top:3px}.markdown-body .task-list-item input{margin:0 .2em .25em -1.6em;vertical-align:middle}.markdown-body hr{border-bottom-color:#eee}.flash{position:relative;padding:16px;color:#246;background-color:#e2eef9;border:1px solid #bac6d3;border-radius:3px}.flash p:last-child{margin-bottom:0}.flash-messages{margin-bottom:24px}.flash-warn{color:#4c4a42;background-color:#fff9ea;border-color:#dfd8c2}.flash-error{color:#911;background-color:#fcdede;border-color:#d2b2b2}.flash-success{color:#22662c;background-color:#e2f9e5;border-color:#bad3be}.flash-plain{color:#4c4a42;background-color:#f5f5f5;border-color:#c1c1c1}</style>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">pre[data-language='renumf90']{background-color:#EFEFFE;}</style>
  <style type="text/css">pre[data-language='blupf90']{background-color:#E7FFE7;}</style>
  <style type="text/css">pre[data-language='output']{background-color:#FFFFFF;border:thin solid black;}</style>
  <style type="text/css">pre[data-language='shell']{background-color:#FFFFFF;border:thin dashed;}</style>
  <style type="text/css">pre[data-caption]::before{content:attr(data-caption) "\a\a";font-size:large;font-weight:bold;border-bottom:thin solid black}</style>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
  </script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<article class="markdown-body">
<header>
<h1 class="title">Data preparation with RENUMF90</h1>
<p class="author">Yutaka Masuda</p>
<p class="date">September 2019</p>
</header>
<nav id="TOC">
<ul>
<li><a href="#advanced-usage-of-renumf90">Advanced usage of RENUMF90</a><ul>
<li><a href="#combined-effects">Combined effects</a></li>
<li><a href="#pedigree-manipulation">Pedigree manipulation</a></li>
<li><a href="#animal-model-options">Animal model options</a></li>
<li><a href="#definition-of-unknown-parent-groups-upgs">Definition of unknown parent groups (UPGs)</a></li>
<li><a href="#considering-inbreeding-coefficients-in-mathbfa-1">Considering inbreeding coefficients in <span class="math inline">\(\mathbf{A}^{-1}\)</span></a></li>
<li><a href="#order-of-keywords">Order of keywords</a></li>
</ul></li>
</ul>
</nav>
<p>
Back to <a href="./index.html">index.html</a>.
</p>
<h1 id="advanced-usage-of-renumf90">Advanced usage of RENUMF90</h1>
<p>In this section, we will introduce some advanced features in RENUMF90. We commonly use the following raw data and pedigree file. The files come from the multiple-trait model shown in the previous chapter.</p>
<pre data-language="text" data-caption="rawdata9.txt"><code>  ID006  A  1  1.0  3.0  4.5
  ID009  A  2  1.0  2.0  7.5
  ID012  A  1  2.0  4.0  3.5
  ID007  B  2  2.0  6.0 -0.5
  ID010  B  1  1.0  3.0  5.5
  ID013  B  2  2.0  6.0  1.5
  ID008  C  1  2.0  6.0 -1.5
  ID011  C  2  1.0  6.0  2.5
  ID014  C  1  1.0  8.0  0.5
  ID015  C  2  2.0  4.0  4.5</code></pre>
<p>The pedigree file is as follows.</p>
<pre data-language="text" data-caption="rawpedigree9.txt"><code> ID001      0      0
 ID002      0      0
 ID003      0      0
 ID004      0      0
 ID005      0      0
 ID006      0      0
 ID007  ID002  ID005
 ID008  ID001  ID004
 ID009  ID002  ID003
 ID010  ID007  ID006
 ID011  ID007  ID004
 ID012  ID011  ID008
 ID013  ID011  ID010
 ID014  ID009  ID013
 ID015  ID011  ID010</code></pre>
<h2 id="combined-effects">Combined effects</h2>
<p>RENUMF90 can treat a field as a combination of the existing fields. This is useful when an interaction effect is incorporated. The keyword <code>COMBINE</code> performs this feature. The following example is from the previous section and the keyword is added.</p>
<pre data-language="renumf90" data-caption="renum9a.txt"><code>COMBINE
6 2 3
DATAFILE
rawdata9.txt
TRAITS
5
FIELDS_PASSED TO OUTPUT

WEIGHT(S)

RESIDUAL_VARIANCE
1.0
EFFECT          # 1st effect
2 cross alpha
EFFECT          # 2nd effect
3 cross alpha
EFFECT          # 3rd effect
4 cov
EFFECT          # interaction effect 1 x 2
6 cross alpha</code></pre>
<p>The <code>COMBINE</code> statement should be used as follows.</p>
<ul>
<li><code>COMBINE</code> must appear at the beginning of the instruction file (before <code>DATAFILE</code>).</li>
<li>At least 2 values are needed; The 2nd or later values will be combined into 1 field. The field can be referred to as the first value.</li>
<li>In this case, <code>6 2 3</code> means that columns 2 and 3 are combined into column 6 (i.e., creating a new column, 6, as a combination of 2 and 3).</li>
<li>You can use any number in the first value even if your data file doesn’t have such a column.</li>
<li>You can repeat the <code>COMBINE</code> statement.</li>
<li>The combined field can be used in the <code>EFFECT</code> statement.</li>
</ul>
<p>You can run RENUMF90 and check <code>renf90.tables</code>. Column 2 and 3 are successfully combined into one effect (group 4).</p>
<pre data-language="text"><code> Effect group 1 of column 1 with 3 levels , effect # 1
 Value    #    consecutive number
A 3 1
B 3 2
C 4 3
 Effect group 2 of column 1 with 2 levels , effect # 2
 Value    #    consecutive number
1 5 1
2 5 2
 Effect group 4 of column 1 with 6 levels , effect # 4
 Value    #    consecutive number
A1 2 1
A2 1 2
B1 1 3
B2 2 4
C1 2 5
C2 2 6</code></pre>
<h2 id="pedigree-manipulation">Pedigree manipulation</h2>
<p>RENUMF90 prunes the pedigree. By default, the program traces back 3-generation (up to great-grandsires and dams) from the animals with phenotype or genotype. The number of generations back can be changed with the option <code>PED_DEPTH</code>, which should place just after <code>SNP_FILE</code> (or <code>FILE_POS</code> if no genetic markers are used). The following statements define 10 generations to be traced.</p>
<pre data-language="renumf90"><code>PED_DEPTH
10</code></pre>
<p>If you put a large number (like 100), you can consider all ancestors to be traced back from the current animals. If you put 0, RENUMF90 tries to include all animals found in the raw pedigree file even if the pedigree animals are not related to the animals with phenotype or genotype.</p>
<h2 id="animal-model-options">Animal model options</h2>
<p>RENUMF90 supports the permanent environmental (PE) effect and the maternal genetic (MG) and the maternal environmental (MPE) effects as additional random effects in the animal model using the <code>OPTIONAL</code> keyword. The following is a parameter file with maximal options (PE, MG, and MPE).</p>
<pre data-language="renumf90" data-caption="renum9b.txt"><code>DATAFILE
rawdata9.txt
TRAITS
5
FIELDS_PASSED TO OUTPUT

WEIGHT(S)

RESIDUAL_VARIANCE
1.0
EFFECT          # 1st effect
2 cross alpha
EFFECT          # 2nd effect
3 cross alpha
EFFECT          # 3rd effect
4 cov
EFFECT          # 4th effect = animal
1 cross alpha
RANDOM
animal
OPTIONAL
pe mat mpe
FILE
rawpedigree9.txt
(CO)VARIANCES
1.0 0.2
0.2 0.3
(CO)VARIANCES_PE
1.5
(CO)VARIANCES_MPE
0.5</code></pre>
<p>You can read this file as follows.</p>
<ul>
<li>The <code>OPTIONAL</code> keyword places just after the <code>EFFECT</code> defining additive genetic effect.</li>
<li>Possible values are <code>pe</code> for the animal’s permanent environmental effect, <code>mat</code> for the maternal genetic effect, and <code>mpe</code> for the maternal environmental effect. They are optional; you can use only the options you want.</li>
<li>If you put <code>mat</code>, you should put a genetic covariance matrix including direct and maternal genetic effects at <code>(CO)VARIANCES</code>. For a single-trait model, the rank of the covariance matrix should be 2; for a <span class="math inline">\(t\)</span> trait model, the rank should be <span class="math inline">\(2t\)</span>. See <em>RENUMF90 Cheat Sheet</em> for details. Without <code>mat</code>, you just put a direct genetic variance (for a single-trait model) or a covariance matrix (for a multiple-trait model).</li>
<li>If you put <code>pe</code>, you should put a permanent-environmental variance (for a single-trait model) or a covariance matrix (for a multiple-trait model) at <code>(CO)VARIANCES_PE</code>. Otherwise, you can omit <code>(CO)VARIANCES_PE</code>.</li>
<li>If you put <code>mpe</code>, you should put a maternal permanent-environmental variance (for a single-trait model) or a covariance matrix (for a multiple-trait model) at <code>(CO)VARIANCES_MPE</code>. Otherwise, you can omit <code>(CO)VARIANCES_MPE</code>.</li>
</ul>
<h2 id="definition-of-unknown-parent-groups-upgs">Definition of unknown parent groups (UPGs)</h2>
<p>RENUMF90 can assign UPGs to unknown parents. There are 2 ways to do this but here we just introduce the one approach (because of its simplicity). The following keyword generates UPGs.</p>
<pre data-language="renumf90"><code>UPG_TYPE
in_pedigrees</code></pre>
<p>The option should be just after <code>PED_DEPTH</code>. With this option, RENUMF90 interprets a negative sire (or dam) ID as a UPG. So you should prepare different columns for sires and dams which contain negative integers for unknown animals instead of 0. Also, all the real animals should be in the pedigree file with the standard ID (other than negative integers).</p>
<p>The following is a simple example of UPGs. The 4th column is for sires with groups (1 and 2) and the 5th column is for dams with groups (3 and 4). You can specify column 4 (instead of 2) as the sire and column 5 (instead of 3) as the dam with the <code>FILE_POS</code> keyword.</p>
<pre data-language="text" data-caption="rawpedigree9c.txt"><code> ID001      0      0     -1     -4
 ID002      0      0     -2     -3
 ID003      0      0     -1     -3
 ID004      0      0     -2     -3
 ID005      0      0     -2     -4
 ID006      0      0     -1     -3
 ID007  ID002  ID005  ID002  ID005
 ID008  ID001  ID004  ID001  ID004
 ID009  ID002  ID003  ID002  ID003
 ID010  ID007  ID006  ID007  ID006
 ID011  ID007  ID004  ID007  ID004
 ID012  ID011  ID008  ID011  ID008
 ID013  ID011  ID010  ID011  ID010
 ID014  ID009  ID013  ID009  ID013
 ID015  ID011  ID010  ID011  ID010</code></pre>
<p>The instruction file can be as follows.</p>
<pre data-language="renumf90" data-caption="renum9c.txt"><code>DATAFILE
rawdata9.txt
TRAITS
5
FIELDS_PASSED TO OUTPUT

WEIGHT(S)

RESIDUAL_VARIANCE
1.0
EFFECT          # 1st effect
2 cross alpha
EFFECT          # 2nd effect
3 cross alpha
EFFECT          # 3rd effect
4 cov
EFFECT          # 4th effect = animal
1 cross alpha
RANDOM
animal
FILE
rawpedigree9c.txt
FILE_POS        #  positions of animal, sire, dam, 0, 0
1 4 5 0 0
UPG_TYPE
in_pedigrees
(CO)VARIANCES
1.0</code></pre>
<p>The resulting pedigree file is as follows. You have only 15 real animals but there are the codes greater than 15 (between 16 and 19) in the sire and dam columns. These numbers greater than the number of real animals represent UPGs.</p>
<pre data-language="text"><code>1 7 5 1 0 2 1 0 0 ID015
13 17 18 3 0 0 0 0 2 ID004
11 17 19 3 0 0 0 0 1 ID005
2 16 18 3 0 0 1 0 1 ID006
3 12 11 1 0 2 1 2 0 ID007
4 14 13 1 0 2 1 0 1 ID008
5 3 2 1 0 2 1 0 2 ID010
6 12 15 1 0 2 1 1 0 ID009
7 3 13 1 0 2 1 3 0 ID011
8 7 4 1 0 2 1 0 0 ID012
14 16 19 3 0 0 0 1 0 ID001
9 7 5 1 0 2 1 0 1 ID013
12 17 18 3 0 0 0 2 0 ID002
10 6 9 1 0 2 1 0 0 ID014
15 16 18 3 0 0 0 0 1 ID003</code></pre>
<h2 id="considering-inbreeding-coefficients-in-mathbfa-1">Considering inbreeding coefficients in <span class="math inline">\(\mathbf{A}^{-1}\)</span></h2>
<p>By default, RENUMF90 creates a pedigree file which will result in <span class="math inline">\(\mathbf{A}^{-1}\)</span> without inbreeding. To put inbreeding coefficients in <span class="math inline">\(\mathbf{A}^{-1}\)</span>, you have to generate a special, renumbered pedigree-file by RENUMF90 using the <code>INBREEDING</code> keyword.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> This keyword should be placed after <code>UPG_TYPE</code>.</p>
<p>The basic usage of this option is shown below.</p>
<pre data-language="renumf90"><code>INBREEDING
pedigree</code></pre>
<p>You should put the literal word <code>pedigree</code> here (you shouldn’t replace it with the pedigree file name). With this option, RENUMF90 calculates inbreeding coefficients with Meuwissen and Luo (1992) based on the pedigree to be saved in the file <code>renaddxx.ped</code> which has been possibly pruned by some generations (see <code>PED_DEPTH</code> for details).</p>
<p>If you want to supply pre-calculated inbreeding coefficients stored in a file, use the different option. Assuming your file is <code>inb.txt</code>, the following option can be used.</p>
<pre data-language="renumf90"><code>INBREEDING
file inb.txt</code></pre>
<p>The file is a space-separated text file with 2 columns: 1) the original animal ID and 2) inbreeding coefficient (ranging from 0 to 1).</p>
<p>The following example uses the pedigree option.</p>
<pre data-language="renumf90" data-caption="renum9d.txt"><code>DATAFILE
rawdata9.txt
TRAITS
5
FIELDS_PASSED TO OUTPUT

WEIGHT(S)

RESIDUAL_VARIANCE
1.0
EFFECT          # 1st effect
2 cross alpha
EFFECT          # 2nd effect
3 cross alpha
EFFECT          # 3rd effect
4 cov
EFFECT          # 4th effect = animal
1 cross alpha
RANDOM
animal
FILE
rawpedigree9.txt
INBREEDING
pedigree
(CO)VARIANCES
1.0</code></pre>
<p>The resulting pedigree file <code>renadd04.ped</code> has the inb/upg code in the 4th column.</p>
<pre data-language="text" data-caption="renadd04.ped"><code>1 7 5 2000 0 2 1 0 0 ID015
13 0 0 1000 0 0 0 0 2 ID004
11 0 0 1000 0 0 0 0 1 ID005
2 0 0 1000 0 0 1 0 1 ID006
3 12 11 2000 0 2 1 2 0 ID007
4 14 13 2000 0 2 1 0 1 ID008
5 3 2 2000 0 2 1 0 2 ID010
6 12 15 2000 0 2 1 1 0 ID009
7 3 13 2000 0 2 1 3 0 ID011
8 7 4 2000 0 2 1 0 0 ID012
14 0 0 1000 0 0 0 1 0 ID001
9 7 5 2000 0 2 1 0 1 ID013
12 0 0 1000 0 0 0 2 0 ID002
10 6 9 2133 0 2 1 0 0 ID014
15 0 0 1000 0 0 0 0 1 ID003</code></pre>
<p>Looking at the 4th column, when an animal has non-inbred parents, the value should be 1000 with unknown parents, 1333 with one parent unknown, or 2000 with known parents. Animal 10 (ID014) has inbred parents so the inb/upg code is different from the other animals.</p>
<h2 id="order-of-keywords">Order of keywords</h2>
<p>The keywords used in the instruction file should be ordered following the manual. Here we show the exact order of keywords supported by RENUMF90.</p>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">Keyword</th>
<th style="text-align: left;">optional</th>
<th style="text-align: left;">possible values</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>COMBINE</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">definition of new field as a combination of existing fields</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>DATAFILE</code></td>
<td style="text-align: left;">mandatory</td>
<td style="text-align: left;">name of raw data file</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>TRAITS</code></td>
<td style="text-align: left;">mandatory</td>
<td style="text-align: left;">positions of observations in the raw data file</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>FIELDS_PASSED</code></td>
<td style="text-align: left;">mandatory</td>
<td style="text-align: left;">positions of items in the raw data file to be passed to <code>renf90.dat</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>WEIGHT(S)</code></td>
<td style="text-align: left;">mandatory</td>
<td style="text-align: left;">positions of weights in the raw data file</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>RESIDUAL_VARIANCE</code></td>
<td style="text-align: left;">mandatory</td>
<td style="text-align: left;">residual covariance matrix</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>EFFECT</code></td>
<td style="text-align: left;">mandatory</td>
<td style="text-align: left;">effect description</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>NESTED</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">positions of nested covariates</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>RANDOM</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">declaration of random effect</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>FILE</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">name of raw pedigree file</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>FILE_POS</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">positions of animal ID, sire ID and dam ID</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>SNP_FILE</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">name of SNP marker file</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>PED_DEPTH</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">the maximum generation back from animals with phenotype or genotype</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>GEN_INT</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">generation interval to set unknown parent groups (UPG)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>REC_SEX</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">check if records are found in specific sex</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>UPG_TYPE</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">UPG specification</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>INBREEDING</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">create pedigree file with inbreeding</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>RANDOM_REGRESSION</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">put covariates for random regressions</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>RR_POSITION</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">positions of covariates for random regressions</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>(CO)VARIANCES</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">covariance components</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>(CO)VARIANCES_PE</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">covariance components for animal’s PE effects</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>(CO)VARIANCES_MPE</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">covariance components for maternal PE effects</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>OPTION</code></td>
<td style="text-align: left;">optional</td>
<td style="text-align: left;">option parameters</td>
</tr>
</tbody>
</table>
<section class="footnotes">
<hr />
<ol>
<li id="fn1"><p>Technically, the alternative, renumbered pedigree-file has the same as the standard one except for the 4th column. In the alternative pedigree, the 4th column has a special 4-digit integer (inb/upg code) which combines the inbreeding coefficients and the number of missing parents. Although the user can manually calculate the code, this process may be complicated, and I recommend the user to use RENUMF90 to generate the pedigree file with such a code.<a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</section>
<p>
Back to <a href="./index.html">index.html</a>.
</p>
</article>
</body>
</html>
