<!DOCTYPE html>
<!--==============================================================================
	           "GitHub HTML5 Pandoc Template" v1.2 — by Tristano Ajmone
	==============================================================================
	(c) Tristano Ajmone, 2017, MIT License (MIT). Project's home repository:

	- https://github.com/tajmone/pandoc-goodies

	This template reuses source code taken from the following projects:

	- GitHub Markdown CSS: © Sindre Sorhus, MIT License (MIT):
	  https://github.com/sindresorhus/github-markdown-css

	- Primer CSS: © 2016 GitHub Inc., MIT License (MIT):
	  http://primercss.io/
	==============================================================================-->
<html>
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Yutaka Masuda" />
  <title>Variance component estimation</title>
  <style type="text/css">@font-face{font-family:octicons-link;src:url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff')}.markdown-body{-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;color:#24292e;font-family:-apple-system,system-ui,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";font-size:16px;line-height:1.5;word-wrap:break-word;box-sizing:border-box;min-width:200px;max-width:980px;margin:0 auto;padding:45px}.markdown-body .octicon{display:inline-block;fill:currentColor;vertical-align:text-bottom}.markdown-body a{background-color:transparent;-webkit-text-decoration-skip:objects;color:#0366d6;text-decoration:none}.markdown-body a:active,.markdown-body a:hover{outline-width:0}.markdown-body h1{margin:.67em 0}.markdown-body img{border-style:none}.markdown-body svg:not(:root){overflow:hidden}.markdown-body code,.markdown-body kbd,.markdown-body pre{font-family:monospace,monospace}.markdown-body input{font:inherit;margin:0;overflow:visible;font-family:inherit;font-size:inherit;line-height:inherit}.markdown-body [type=checkbox]{box-sizing:border-box;padding:0}.markdown-body *{box-sizing:border-box}.markdown-body a:hover{text-decoration:underline}.markdown-body strong{font-weight:600}.markdown-body hr{box-sizing:content-box;overflow:hidden;background:0 0;border-bottom:1px solid #dfe2e5}.markdown-body hr::before{display:table;content:""}.markdown-body hr::after{display:table;clear:both;content:""}.markdown-body table{border-spacing:0;border-collapse:collapse;display:block;width:100%;overflow:auto}.markdown-body td,.markdown-body th{padding:0}.markdown-body blockquote{margin:0}.markdown-body ol ol,.markdown-body ul ol{list-style-type:lower-roman}.markdown-body ol ol ol,.markdown-body ol ul ol,.markdown-body ul ol ol,.markdown-body ul ul ol{list-style-type:lower-alpha}.markdown-body dd{margin-left:0}.markdown-body code{font-family:SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace}.markdown-body pre{font:12px SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace;word-wrap:normal}.markdown-body .pl-0{padding-left:0!important}.markdown-body .pl-1{padding-left:4px!important}.markdown-body .pl-2{padding-left:8px!important}.markdown-body .pl-3{padding-left:16px!important}.markdown-body .pl-4{padding-left:24px!important}.markdown-body .pl-5{padding-left:32px!important}.markdown-body .pl-6{padding-left:40px!important}.markdown-body::before{display:table;content:""}.markdown-body::after{display:table;clear:both;content:""}.markdown-body>:first-child{margin-top:0!important}.markdown-body>:last-child{margin-bottom:0!important}.markdown-body a:not([href]){color:inherit;text-decoration:none}.markdown-body .anchor{float:left;padding-right:4px;margin-left:-20px;line-height:1}.markdown-body .anchor:focus{outline:0}.markdown-body blockquote,.markdown-body dl,.markdown-body ol,.markdown-body p,.markdown-body pre,.markdown-body table,.markdown-body ul{margin-top:0;margin-bottom:16px}.markdown-body hr{height:.25em;padding:0;margin:24px 0;background-color:#e1e4e8;border:0}.markdown-body blockquote{padding:0 1em;color:#6a737d;border-left:.25em solid #dfe2e5}.markdown-body blockquote>:first-child{margin-top:0}.markdown-body blockquote>:last-child{margin-bottom:0}.markdown-body kbd{font-size:11px;box-shadow:inset 0 -1px 0 #959da5}.markdown-body h1,.markdown-body h2,.markdown-body h3,.markdown-body h4,.markdown-body h5,.markdown-body h6{margin-top:24px;margin-bottom:16px;font-weight:600;line-height:1.25}.markdown-body h1 .octicon-link,.markdown-body h2 .octicon-link,.markdown-body h3 .octicon-link,.markdown-body h4 .octicon-link,.markdown-body h5 .octicon-link,.markdown-body h6 .octicon-link{color:#1b1f23;vertical-align:middle;visibility:hidden}.markdown-body h1:hover .anchor,.markdown-body h2:hover .anchor,.markdown-body h3:hover .anchor,.markdown-body h4:hover .anchor,.markdown-body h5:hover .anchor,.markdown-body h6:hover .anchor{text-decoration:none}.markdown-body h1:hover .anchor .octicon-link,.markdown-body h2:hover .anchor .octicon-link,.markdown-body h3:hover .anchor .octicon-link,.markdown-body h4:hover .anchor .octicon-link,.markdown-body h5:hover .anchor .octicon-link,.markdown-body h6:hover .anchor .octicon-link{visibility:visible}.markdown-body h1{padding-bottom:.3em;font-size:2em;border-bottom:1px solid #eaecef}.markdown-body h2{padding-bottom:.3em;font-size:1.5em;border-bottom:1px solid #eaecef}.markdown-body h3{font-size:1.25em}.markdown-body h4{font-size:1em}.markdown-body h5{font-size:.875em}.markdown-body h6{font-size:.85em;color:#6a737d}.markdown-body ol,.markdown-body ul{padding-left:2em}.markdown-body ol ol,.markdown-body ol ul,.markdown-body ul ol,.markdown-body ul ul{margin-top:0;margin-bottom:0}.markdown-body li>p{margin-top:16px}.markdown-body li+li{margin-top:.25em}.markdown-body dl{padding:0}.markdown-body dl dt{padding:0;margin-top:16px;font-size:1em;font-style:italic;font-weight:600}.markdown-body dl dd{padding:0 16px;margin-bottom:16px}.markdown-body table th{font-weight:600}.markdown-body table td,.markdown-body table th{padding:6px 13px;border:1px solid #dfe2e5}.markdown-body table tr{background-color:#fff;border-top:1px solid #c6cbd1}.markdown-body table tr:nth-child(2n){background-color:#f6f8fa}.markdown-body img{max-width:100%;box-sizing:content-box;background-color:#fff}.markdown-body code{padding:.2em 0;margin:0;font-size:85%;background-color:rgba(27,31,35,.05);border-radius:3px}.markdown-body code::after,.markdown-body code::before{letter-spacing:-.2em;content:"\00a0"}.markdown-body pre>code{padding:0;margin:0;font-size:100%;word-break:normal;white-space:pre;background:0 0;border:0}.markdown-body .highlight{margin-bottom:16px}.markdown-body .highlight pre{margin-bottom:0;word-break:normal}.markdown-body .highlight pre,.markdown-body pre{padding:16px;overflow:auto;font-size:85%;line-height:1.45;background-color:#f6f8fa;border-radius:3px}.markdown-body pre code{display:inline;max-width:auto;padding:0;margin:0;overflow:visible;line-height:inherit;word-wrap:normal;background-color:transparent;border:0}.markdown-body pre code::after,.markdown-body pre code::before{content:normal}.markdown-body .full-commit .btn-outline:not(:disabled):hover{color:#005cc5;border-color:#005cc5}.markdown-body kbd{display:inline-block;padding:3px 5px;font:11px SFMono-Regular,Consolas,"Liberation Mono",Menlo,Courier,monospace;line-height:10px;color:#444d56;vertical-align:middle;background-color:#fcfcfc;border:1px solid #c6cbd1;border-bottom-color:#959da5;border-radius:3px;box-shadow:inset 0 -1px 0 #959da5}.markdown-body :checked+.radio-label{position:relative;z-index:1;border-color:#0366d6}.markdown-body .task-list-item{list-style-type:none}.markdown-body .task-list-item+.task-list-item{margin-top:3px}.markdown-body .task-list-item input{margin:0 .2em .25em -1.6em;vertical-align:middle}.markdown-body hr{border-bottom-color:#eee}.flash{position:relative;padding:16px;color:#246;background-color:#e2eef9;border:1px solid #bac6d3;border-radius:3px}.flash p:last-child{margin-bottom:0}.flash-messages{margin-bottom:24px}.flash-warn{color:#4c4a42;background-color:#fff9ea;border-color:#dfd8c2}.flash-error{color:#911;background-color:#fcdede;border-color:#d2b2b2}.flash-success{color:#22662c;background-color:#e2f9e5;border-color:#bad3be}.flash-plain{color:#4c4a42;background-color:#f5f5f5;border-color:#c1c1c1}</style>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">pre[data-language='renumf90']{background-color:#EFEFFE;}</style>
  <style type="text/css">pre[data-language='blupf90']{background-color:#E7FFE7;}</style>
  <style type="text/css">pre[data-language='output']{background-color:#FFFFFF;border:thin solid black;}</style>
  <style type="text/css">pre[data-language='shell']{background-color:#FFFFFF;border:thin dashed;}</style>
  <style type="text/css">pre[data-caption]::before{content:attr(data-caption) "\a\a";font-size:large;font-weight:bold;border-bottom:thin solid black}</style>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
  </script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<article class="markdown-body">
<header>
<h1 class="title">Variance component estimation</h1>
<p class="author">Yutaka Masuda</p>
<p class="date">September 2019</p>
</header>
<nav id="TOC">
<ul>
<li><a href="#advanced-usage-of-airemlf90">Advanced usage of AIREMLF90</a><ul>
<li><a href="#heterogeneous-residual-variances">Heterogeneous residual variances</a></li>
<li><a href="#likelihood-ratio-test">Likelihood Ratio Test</a></li>
</ul></li>
</ul>
</nav>
<h1 id="advanced-usage-of-airemlf90">Advanced usage of AIREMLF90</h1>
<h2 id="heterogeneous-residual-variances">Heterogeneous residual variances</h2>
<p>There is a situation where the residual variances vary over conditions. This is known as heterogeneous residual variances. There are 2 types of modeling for the heterogeneous residual variances in AIREMLF90.</p>
<ol type="1">
<li>The residual variance is related to a covariate. For example, the residual variance for body weight increase by age of a month. The residual variance can be described as a function of a covariate (age).</li>
<li>The residual variance differs by class. For example, in a test-day random-regression model, we usually assume different residual variance in each lactation stage. The residual variances are independent of each other.</li>
</ol>
<p>Although AIREMLF90 is designed to handle the first case, the program can also handle the second case with a trick. We will see both cases with a numerical example in REML. GIBBS3F90 (and THRGIBBS3F90) supports the second case.</p>
<h3 id="model">Model</h3>
<p>In this example, we asuume the following model: <span class="math display">\[
y_{ijkl} = F_{i} + S_{j} +Y_{k} + u_{l} + e_{ijkl}
\]</span> where <span class="math inline">\(y_{ijkl}\)</span> is the observation, <span class="math inline">\(F_{i}\)</span> , <span class="math inline">\(S_j\)</span>, and <span class="math inline">\(Y_k\)</span> are the fixed effects, <span class="math inline">\(u_l\)</span> is the additive genetic effect and <span class="math inline">\(e_{ijkl}\)</span> is the residual effect. The residual variance is defined as the following function <span class="math display">\[
\sigma_e^2 = \exp(b_0 + b_1 x_{ijkl})
\]</span> where <span class="math inline">\(b_0\)</span> and <span class="math inline">\(b_1\)</span> are the regression coefficients to account for the residual variance and <span class="math inline">\(x_{ijkl}\)</span> is the covariate measured with the observation <span class="math inline">\(y_{ijkl}\)</span>. The model is based on Foulley and Quaas (1995). AIREMLF90 will estimate <span class="math inline">\(b_0\)</span> and <span class="math inline">\(b_1\)</span> with user-supplied <span class="math inline">\(x_{ijkl}\)</span>.</p>
<h3 id="data">Data</h3>
<p>In this section, we will use simulated files similar to the previous section. You can download the files from Github.</p>
<ul>
<li><a href="https://github.com/Masuday/data/blob/master/tutorial/simdata2.txt"><code>simdata2.txt</code></a> : data file</li>
<li><a href="https://github.com/Masuday/data/blob/master/tutorial/simped2.txt"><code>simped2.txt</code></a> : pedigree file</li>
</ul>
<p>The pedigree file contains 3 columns: animal, sire, and dam. The data file has 12 columns as described below.</p>
<table>
<thead>
<tr class="header">
<th>Column</th>
<th style="text-align: left;">Item</th>
<th style="text-align: left;">type</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td style="text-align: left;">Animal ID</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">Same as in pedigree (4641 animals)</td>
</tr>
<tr class="even">
<td>2</td>
<td style="text-align: left;">Sire ID</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">Same as in pedigree</td>
</tr>
<tr class="odd">
<td>3</td>
<td style="text-align: left;">Dam ID</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">Same as in pedigree</td>
</tr>
<tr class="even">
<td>4</td>
<td style="text-align: left;">Weight</td>
<td style="text-align: left;">real</td>
<td style="text-align: left;">Not used here</td>
</tr>
<tr class="odd">
<td>5</td>
<td style="text-align: left;">Mu</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">All 1: not used here</td>
</tr>
<tr class="even">
<td>6</td>
<td style="text-align: left;">Farm</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">Class effect (155 levels)</td>
</tr>
<tr class="odd">
<td>7</td>
<td style="text-align: left;">Sex</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">Class effect (2 levels)</td>
</tr>
<tr class="even">
<td>8</td>
<td style="text-align: left;">Year</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">Class effect (11 levels)</td>
</tr>
<tr class="odd">
<td>9</td>
<td style="text-align: left;">Obs. 1</td>
<td style="text-align: left;">real</td>
<td style="text-align: left;">Phenotype for trait 1</td>
</tr>
<tr class="even">
<td>10</td>
<td style="text-align: left;">Obs. 2</td>
<td style="text-align: left;">real</td>
<td style="text-align: left;">Phenotype for trait 2</td>
</tr>
<tr class="odd">
<td>11</td>
<td style="text-align: left;">Obs. 3</td>
<td style="text-align: left;">real</td>
<td style="text-align: left;">Phenotype for trait 3</td>
</tr>
<tr class="even">
<td>12</td>
<td style="text-align: left;">Obs. 4</td>
<td style="text-align: left;">real</td>
<td style="text-align: left;">Phenotype for trait 4</td>
</tr>
<tr class="odd">
<td>13</td>
<td style="text-align: left;">Covariate</td>
<td style="text-align: left;">real</td>
<td style="text-align: left;">Related to residual variance</td>
</tr>
<tr class="even">
<td>14</td>
<td style="text-align: left;">Class</td>
<td style="text-align: left;">integer</td>
<td style="text-align: left;">(Used in the next subsection)</td>
</tr>
</tbody>
</table>
<p>Column 13 contains a real value as a covariate. In this artificial data, the residual variance is expected to be larger as the covariate becomes larger. The 14th column contains the heterogeneous-residual-variance class (3 levels), and it will be used in the next subsection.</p>
<h3 id="parameter-file">Parameter file</h3>
<p>The following parameter file is used.</p>
<pre data-language="blupf90" data-caption="aireml2.txt"><code>DATAFILE
simdata2.txt
NUMBER_OF_TRAITS
1
NUMBER_OF_EFFECTS
4
OBSERVATION(S)
9
WEIGHT(S)
4
EFFECTS:
 6  155 cross
 7    2 cross
 8   11 cross
 1 4641 cross
RANDOM_RESIDUAL VALUES
100
RANDOM_GROUP
4
RANDOM_TYPE
add_animal
FILE
simped2.txt
(CO)VARIANCES
100
OPTION hetres_pos 13
OPTION hetres_pol 4.5 0.5</code></pre>
<p>The options define the model for heterogeneous residual variances.</p>
<ul>
<li><code>OPTION hetres_pos</code> = the position of covariate <span class="math inline">\(x\)</span> in the data file. Intercept <span class="math inline">\(b_0\)</span> is implicitly considered in the program.</li>
<li><code>OPTION hetres_pol</code> = the starting values for <span class="math inline">\(b_{0}\)</span> and <span class="math inline">\(b_{1}\)</span>.
<ul>
<li>You should supply the values for all coefficients including intercept.</li>
<li>In this example, the initial value will be <span class="math inline">\(\exp(4.5 + 0.5) = 148.4\)</span> for <span class="math inline">\(x = 1\)</span>.</li>
</ul></li>
</ul>
<p>In the above case, we define only 1 regression coefficient in <code>hetres_pos</code> but the program adds the intercept. If you put 2 numbers in <code>hetres_pos</code> and also put 2 initial values in <code>hetres_pol</code>, the program doesn’t add the intercept. In such a case, the model doesn’t contain any intercept but does contain user-supplied covariates only.</p>
<h3 id="results">Results</h3>
<p>You can run AIREMLF90 with the above parameter file. The following output will be shown.</p>
<pre data-language="output"><code>Final Estimates
 Genetic variance(s) for effect  4
   39.384
 new R
           1 -th trait:           1 -th coefficient =   4.25520168650538
           1 -th trait:           2 -th coefficient =  5.886409914144559E-002
 inverse of AI matrix (Sampling Variance)
   16.773     -0.12514      0.77591E-02
 -0.12514      0.41370E-02 -0.16778E-02
  0.77591E-02 -0.16778E-02  0.10722E-02
 Correlations from inverse of AI matrix
   1.0000     -0.47507      0.57858E-01
 -0.47507       1.0000     -0.79664
  0.57858E-01 -0.79664       1.0000
 SE for G
   4.0955
 SE for R
  0.64319E-01</code></pre>
<p>The output shows <span class="math inline">\(\hat{b}_0 = 4.2552\)</span> and <span class="math inline">\(\hat{b}_1 = 0.058864\)</span>. This corresponds to <span class="math inline">\(\hat{\sigma}_e^2 = \exp(4.26 + 0.0589 \times 0.5) = 72.9\)</span> for <span class="math inline">\(x = 0.5\)</span>, <span class="math inline">\(\hat{\sigma}_e^2 = \exp(4.26 + 0.0589 \times 1.5) = 77.4\)</span> for <span class="math inline">\(x = 1.5\)</span> and <span class="math inline">\(\sigma_e^2 = \exp(4.26 + 0.0589 \times 2.5) = 82.0\)</span> for <span class="math inline">\(x = 2.5\)</span>.</p>
<p>One drawback of this analysis is that the calculations of the standard error for the residual variance are not easy. The AI matrix contains information for <span class="math inline">\(\hat{b}_0\)</span> and <span class="math inline">\(\hat{b}_1\)</span> (the last 2 rows/columns).</p>
<h3 id="another-modeling-by-class">Another modeling by class</h3>
<p>The heterogeneous residual variances are accounted for by class. We assume 3 levels in the class. The 14th column in the above data contains the level for each observation. AIREMLF90 doesn’t directly accept the level number, and we should convert it to a different format.</p>
<p>Having 3 levels is equivalent to defining 3 regression coefficients without intercept. We now assume the following parameterization without intercept. <span class="math display">\[
\sigma_e^2 = \exp(b_1 x_1 + b_2 x_2 + b_3 x_3 )
\]</span></p>
<p>If the observation has the residual variance level 1, we have <span class="math inline">\(x_1 = 1\)</span>, <span class="math inline">\(x_2 = 0\)</span>, and <span class="math inline">\(x_3 = 0\)</span>. For the level 2, <span class="math inline">\(x_1 = 0\)</span>, <span class="math inline">\(x_2 = 1\)</span>, and <span class="math inline">\(x_3 = 0\)</span> and for the level 3, <span class="math inline">\(x_1 = 0\)</span>, <span class="math inline">\(x_2 = 0\)</span>, and <span class="math inline">\(x_3 = 1\)</span>. We need 3 covariate with 1 or 0. We add 3 extra columns to the rightmost of the data file.</p>
<ul>
<li>tutorials:<code>simdata2a.txt</code>: extended data file</li>
</ul>
<p>The first 10 rows are shown below.</p>
<pre data-language="text" data-caption="simdata2a.txt"><code>  1       0       0 1.00   1    67   1   1    85.0     0.0    92.1    91.0  0.46  1   1 0 0
  2       0       0 0.98   1   144   1   1   115.2   103.9    94.8    90.6  1.52  2   0 1 0
  3       0       0 1.04   1    92   1   1    93.0   107.4   114.9   107.0  0.00  1   1 0 0
  4       0       0 0.97   1    26   1   1    84.0    91.8    93.7   107.1  0.61  1   1 0 0
  5       0       0 0.96   1    83   1   1   100.9    87.9    88.9    91.5  0.55  1   1 0 0
  6       0       0 0.99   1    62   1   1    69.8    71.5    69.3     0.0  0.88  1   1 0 0
  7       0       0 0.97   1    40   1   1   101.2     0.0    92.1    88.6  0.27  1   1 0 0
  8       0       0 1.07   1    82   1   1    77.4    72.7    84.1   105.2  1.69  2   0 1 0
  9       0       0 1.01   1    71   1   1   107.8    87.0    97.8    76.0  2.89  3   0 0 1
 10       0       0 0.95   1   114   1   1    94.4   100.5   103.8    89.8  1.26  2   0 1 0</code></pre>
<p>The column 15, 16 and 17 contains 1 or 0. In column 15, it is 1 if the level is 1, otherwise 0. In column 16, it is 1 if the level is 2, otherwise 0. In column 17, it is 1 if the level is 3, otherwise 0. With this data file, the parameter file is as follows.</p>
<pre data-language="blupf90" data-caption="aireml2a.txt"><code>DATAFILE
simdata2a.txt
NUMBER_OF_TRAITS
1
NUMBER_OF_EFFECTS
4
OBSERVATION(S)
9
WEIGHT(S)
4
EFFECTS:
 6   155 cross
 7     2 cross
 8    11 cross
 1  4641 cross
RANDOM_RESIDUAL VALUES
100
RANDOM_GROUP
4
RANDOM_TYPE
add_animal
FILE
simped2.txt
(CO)VARIANCES
100
OPTION hetres_pos 15 16 17
OPTION hetres_pol 4.5 4.5 4.5</code></pre>
<p>The results from the above parameter file are shown here.</p>
<pre data-language="output"><code>Final Estimates
 Genetic variance(s) for effect  4
   39.296
 new R
           1 -th trait:           1 -th coefficient =   4.28231169726656
           1 -th trait:           2 -th coefficient =   4.36065608258819
           1 -th trait:           3 -th coefficient =   4.38963108448880
 inverse of AI matrix (Sampling Variance)
   16.735     -0.11982     -0.11288     -0.10685
 -0.11982      0.32743E-02  0.75391E-03  0.69459E-03
 -0.11288      0.75391E-03  0.31352E-02  0.67693E-03
 -0.10685      0.69459E-03  0.67693E-03  0.29536E-02
 Correlations from inverse of AI matrix
   1.0000     -0.51184     -0.49281     -0.48059
 -0.51184       1.0000      0.23530      0.22335
 -0.49281      0.23530       1.0000      0.22245
 -0.48059      0.22335      0.22245       1.0000
 SE for G
   4.0909
 SE for R
  0.57221E-01</code></pre>
<p>The estimate for the level 1 is <span class="math inline">\(\hat{\sigma}_e^2 = \exp(4.28) = 72.2\)</span>, for the level 2, <span class="math inline">\(\hat{\sigma}_e^2 = \exp(4.36) = 78.3\)</span> and for the level 3, <span class="math inline">\(\hat{\sigma}_e^2 = \exp(4.39) = 80.6\)</span>. These values are very similar to the previous results from a regression parameterization i.e. <span class="math inline">\(\exp(b_0 + b_1 x)\)</span> for <span class="math inline">\(x = 0.5\)</span>, <span class="math inline">\(1.5\)</span>, and <span class="math inline">\(2.5\)</span>.</p>
<h3 id="remarks">Remarks</h3>
<p>The methods explained here will not work well for multiple-trait models or complicated models with many variance components.</p>
<h2 id="likelihood-ratio-test">Likelihood Ratio Test</h2>
<p>REML methods allow statistical testing of variance components. For instance, imagine that we have to choose one of two competing models:</p>
<ul>
<li>model 1 <em>not</em> including the maternal permanent environmental effect</li>
<li>model 2 <em>including</em> maternal permanent effects.</li>
</ul>
<p>The Likelihood Ratio Test (LRT) checks if the extra random effect, with associated variance component gives a better fit of the model, against not fitting it. The model <em>without</em> the extra random effect is the <em>null</em> (<span class="math inline">\(H_0\)</span>) model versus the model <em>with</em> the extra random effect, which is the alternative model (<span class="math inline">\(H_1\)</span>). Then the difference between the two, which is a positive number, is an LRT statistic which follows a mixture of <span class="math inline">\(\chi^2\)</span> distributions. The theory of the LRT can be found in standard books and a nice description is in Sorensen &amp; Gianola book.</p>
<p>Output of AIREMLF90 under <span class="math inline">\(H_0\)</span> is <span class="math inline">\(x=-2logL\)</span> ; output under <span class="math inline">\(H_1\)</span> is <span class="math inline">\(y=-2logL\)</span>. These are positive numbers, so the smaller the better; always <span class="math inline">\(y&lt;x\)</span> as <span class="math inline">\(H_1\)</span> is a more complex (so more likely) model. The trick resides in knowing if <span class="math inline">\(x-y\)</span> is “big enough”. The LRT statistic is <span class="math inline">\(LRT=x-y\)</span>, which is a statistic distributed as <span class="math inline">\(\chi^2\)</span>. For one variance component, the p-value is (in R) :</p>
<pre data-language="r"><code>pchisq(x-y,1,lower.tail=FALSE)/2</code></pre>
<p>Classical applications of LRT in quantitative genetics include testing <span class="math inline">\(h^2 &gt; 0\)</span> or testing of QTL effects. An application for association analysis for a multi-allelic gene is (<a href="http://dx.doi.org/10.3168/jds.2013-6570" class="uri">http://dx.doi.org/10.3168/jds.2013-6570</a>).</p>
<h3 id="example-for-test-of-heritability-in-dairy-sheep">Example for test of heritability in dairy sheep</h3>
<p>I tested if genetic effects (<span class="math inline">\(\mathbf{u}\)</span>) should be fit in an old data set from dairy sheep (80,000 records, 50,000 animals in pedigree) with a model including permanent (<span class="math inline">\(\sigma^2_p\)</span>), genetic (<span class="math inline">\(\sigma^2_p\)</span>) and residual (<span class="math inline">\(\sigma^2_e\)</span>) variances.</p>
<p>Under the complete model H1 (<span class="math inline">\(\mathbf{y=Xb+Zu+Zp+e}\)</span>) ) we have in the (last lines of) the output of AIREMLF90 :</p>
<pre><code>-2logL =        825227.48</code></pre>
<p>Then we run the reduced model H0 with a parameter file that will <em>not</em> include the genetic effect (<span class="math inline">\(\mathbf{y=Xb+Zp+e}\)</span>) we have:</p>
<pre><code> -2logL =    828095.614456413</code></pre>
<p>then the test in R is:</p>
<pre data-language="R"><code>pchisq(828095.614456413-825227.48,1,lower.tail=FALSE)/2

[1] 0</code></pre>
<p>the p-value is so small that we cannot see the difference from 0. We can take the log, then transform to a scale of <span class="math inline">\(-\log_{10}(\mathrm{p-value})\)</span>, a scale typically used in GWAS:</p>
<pre data-language="R"><code>pval=pchisq(828095.614456413-825227.48,1,lower.tail=FALSE,log.p=TRUE)/2
&gt; pval
[1] -719.137
-log10(exp(pval))
[1] 312.3172</code></pre>
<p>So, the p-value is <span class="math inline">\(p&lt;10^{-312}\)</span>, the evidence against <span class="math inline">\(H_0\)</span> is very strong and we should accept that there is genetic variance in this data set.</p>
</article>
</body>
</html>
